Expose private APIs to enable/disable pickling ScriptModules without RPC

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39631 Expose private APIs to enable/disable pickling ScriptModules without RPC**

Background:
Currently, we cannot send ScriptModule over RPC as an argument.
Otherwise, it would hit the following error:

> _pickle.PickleError: ScriptModules cannot be deepcopied using
> copy.deepcopy or saved using torch.save. Mixed serialization of
> script and non-script modules is not supported. For purely
> script modules use my_script_module.save(<filename>) instead.

Failed attempt:
tried to install `torch.jit.ScriptModule` to RPC's
dispatch table, but it does not work as the dispatch table only
matches exact types and using base type `torch.jit.ScriptModule`
does not work for derived typed.

Current solution:
The current solution exposes `_enable_jit_rref_pickle` and
`_disable_jit_rref_pickle` APIs to toggle the `allowJitRRefPickle`
flag. See `test_pickle_script_module_with_rref` as an example.

Differential Revision: [D21920870](https://our.internmc.facebook.com/intern/diff/D21920870)

mrshenli committed:
Expose private APIs to enable/disable pickling ScriptModules without RPC

Failed attempt: tried to install `torch.jit.ScriptModule` to RPC's
dispatch table, but it does not work as the dispatch table only
matches exact types and using base type `torch.jit.ScriptModule`
does not work for derived typed.

The current solution exposes `_enable_jit_rref_pickle` and
`_disable_jit_rref_pickle` APIs to toggle the `allowJitRRefPickle`
flag. See `test_pickle_script_module_with_rref` as an example.

[ghstack-poisoned]

mrshenli committed:
Update on "Expose private APIs to enable/disable pickling ScriptModules without RPC"


Failed attempt: tried to install `torch.jit.ScriptModule` to RPC's
dispatch table, but it does not work as the dispatch table only
matches exact types and using base type `torch.jit.ScriptModule`
does not work for derived typed.

The current solution exposes `_enable_jit_rref_pickle` and
`_disable_jit_rref_pickle` APIs to toggle the `allowJitRRefPickle`
flag. See `test_pickle_script_module_with_rref` as an example.

[ghstack-poisoned]

mrshenli committed:
Update on "Expose private APIs to enable/disable pickling ScriptModules without RPC"


Background:
Currently, we cannot send ScriptModule over RPC as an argument.
Otherwise, it would hit the following error:

> _pickle.PickleError: ScriptModules cannot be deepcopied using
> copy.deepcopy or saved using torch.save. Mixed serialization of
> script and non-script modules is not supported. For purely
> script modules use my_script_module.save(<filename>) instead.

Failed attempt:
tried to install `torch.jit.ScriptModule` to RPC's
dispatch table, but it does not work as the dispatch table only
matches exact types and using base type `torch.jit.ScriptModule`
does not work for derived typed.

Current solution:
The current solution exposes `_enable_jit_rref_pickle` and
`_disable_jit_rref_pickle` APIs to toggle the `allowJitRRefPickle`
flag. See `test_pickle_script_module_with_rref` as an example.

[ghstack-poisoned]

wanchaol commented on torch/testing/_internal/distributed/rpc/jit/rpc_test.py:
```diff
@@ -920,6 +927,29 @@ def test_remote_script_module(self):
+        import pickle
+
+        f = io.BytesIO()
+        p = pickle.Pickler(f)
```
is the pickler getting used anywhere?

wanchaol commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -693,6 +693,25 @@ PyObject* rpc_init(PyObject* /* unused */) {
+      "_enable_jit_rref_pickle",
+      &enableJitRRefPickle,
+      R"(
+        Allows ``torch.jit.save`` to pickle a ``torch.jit.ScriptModule`` out of
```
seems like we are using torch.jit.save instead of `pickle` to save the ScriptModule, the doc here should be more clear on that I think.

We should probably say:
```
Allows ``torch.jit.save`` to save a ``torch.jit.ScriptModule`` with pickled RRefs out of RPC contexts.
```

wanchaol commented on torch/testing/_internal/distributed/rpc/jit/rpc_test.py:
```diff
@@ -832,6 +833,12 @@ def async_wrong_type():
     return torch.zeros(2)
 
 
+def run_pickled_script_module(pickled_script_module):
```
nit: name probably should be `load_script_module_with_pickled_rref`

wanchaol commented on torch/testing/_internal/distributed/rpc/jit/rpc_test.py:
```diff
@@ -920,6 +927,29 @@ def test_remote_script_module(self):
             )
 
+    @dist_init
+    def test_pickle_script_module_with_rref(self):
```
ditto

wanchaol reviewed (approved):
looks good, please see inline comments.

mrshenli committed:
Update on "Expose private APIs to enable/disable pickling ScriptModules without RPC"


Background:
Currently, we cannot send ScriptModule over RPC as an argument.
Otherwise, it would hit the following error:

> _pickle.PickleError: ScriptModules cannot be deepcopied using
> copy.deepcopy or saved using torch.save. Mixed serialization of
> script and non-script modules is not supported. For purely
> script modules use my_script_module.save(<filename>) instead.

Failed attempt:
tried to install `torch.jit.ScriptModule` to RPC's
dispatch table, but it does not work as the dispatch table only
matches exact types and using base type `torch.jit.ScriptModule`
does not work for derived typed.

Current solution:
The current solution exposes `_enable_jit_rref_pickle` and
`_disable_jit_rref_pickle` APIs to toggle the `allowJitRRefPickle`
flag. See `test_pickle_script_module_with_rref` as an example.

Differential Revision: [D21920870](https://our.internmc.facebook.com/intern/diff/D21920870)

[ghstack-poisoned]

mrshenli committed:
Update on "Expose private APIs to enable/disable pickling ScriptModules without RPC"


Background:
Currently, we cannot send ScriptModule over RPC as an argument.
Otherwise, it would hit the following error:

> _pickle.PickleError: ScriptModules cannot be deepcopied using
> copy.deepcopy or saved using torch.save. Mixed serialization of
> script and non-script modules is not supported. For purely
> script modules use my_script_module.save(<filename>) instead.

Failed attempt:
tried to install `torch.jit.ScriptModule` to RPC's
dispatch table, but it does not work as the dispatch table only
matches exact types and using base type `torch.jit.ScriptModule`
does not work for derived typed.

Current solution:
The current solution exposes `_enable_jit_rref_pickle` and
`_disable_jit_rref_pickle` APIs to toggle the `allowJitRRefPickle`
flag. See `test_pickle_script_module_with_rref` as an example.

Differential Revision: [D21920870](https://our.internmc.facebook.com/intern/diff/D21920870)

[ghstack-poisoned]

mrshenli committed:
Update on "Expose private APIs to enable/disable pickling ScriptModules without RPC"


Background:
Currently, we cannot send ScriptModule over RPC as an argument.
Otherwise, it would hit the following error:

> _pickle.PickleError: ScriptModules cannot be deepcopied using
> copy.deepcopy or saved using torch.save. Mixed serialization of
> script and non-script modules is not supported. For purely
> script modules use my_script_module.save(<filename>) instead.

Failed attempt:
tried to install `torch.jit.ScriptModule` to RPC's
dispatch table, but it does not work as the dispatch table only
matches exact types and using base type `torch.jit.ScriptModule`
does not work for derived typed.

Current solution:
The current solution exposes `_enable_jit_rref_pickle` and
`_disable_jit_rref_pickle` APIs to toggle the `allowJitRRefPickle`
flag. See `test_pickle_script_module_with_rref` as an example.

Differential Revision: [D21920870](https://our.internmc.facebook.com/intern/diff/D21920870)

[ghstack-poisoned]
